osquery/remote/http_client.cpp

```diff
@@ -412,23 +413,44 @@ Response Client::sendHTTPRequest(Request& req) {
           return Response(resp.release());
         }
 
+        if (redirect_attempts++ >= 10) {
+          throw std::runtime_error("Exceeded max of 10 redirects");
+        }
+
         std::string redir_url = Response(resp.release()).headers()["Location"];
         if (!redir_url.size()) {
           throw std::runtime_error(
               "Location header missing in redirect response");
         }
 
-        req.uri(redir_url);
         VLOG(1) << "HTTP(S) request re-directed to: " << redir_url;
+        if (redir_url[0] == '/') {
+          // Relative URI.
+          if (req.remotePort()) {
+            redir_url.insert(0, *req.remotePort());
+            redir_url.insert(0, ":");
+          }
+          if (req.remoteHost()) {
+            redir_url.insert(0, *req.remoteHost());
+          }
+          if (req.protocol()) {
+            redir_url.insert(0, "://");
+            redir_url.insert(0, *req.protocol());
+          }
+        } else {
+          // Absolute URI.
+          init_request = true;
```

Comment:
A possible optimization would be to check whether the protocol, host, and port match before setting this and initializing a new connection. Not necessary IMO for merging this PR.